alacritty/src/wayland_theme.rs

```diff
@@ -32,51 +35,53 @@ impl AlacrittyWaylandTheme {
-            ButtonState::Disabled => {
-                [0xff, self.dim_foreground.r, self.dim_foreground.g, self.dim_foreground.b]
-            },
+    fn button_foreground_color(&self, color: Rgb, status: ButtonState, window_active: bool) -> Rgb {
+        match (window_active, status) {
+            (false, _) => self.dim_foreground,
+            (_, ButtonState::Hovered) => color,
+            (_, ButtonState::Idle) => self.foreground,
+            (_, ButtonState::Disabled) => self.dim_foreground,
         }
     }
 }
 
 impl WaylandTheme for AlacrittyWaylandTheme {
-    fn primary_color(&self, _window_active: bool) -> [u8; 4] {
-        [0xff, self.background.r, self.background.g, self.background.b]
-    }
-
-    fn secondary_color(&self, window_active: bool) -> [u8; 4] {
-        if window_active {
-            [0xff, self.foreground.r, self.foreground.g, self.foreground.b]
-        } else {
-            [0xff, self.dim_foreground.r, self.dim_foreground.g, self.dim_foreground.b]
+    fn element_color(&self, element: Element, window_active: bool) -> ARGBColor {
+        match element {
+            Element::Bar | Element::Separator => self.background.into_rgba(),
+            Element::Text if window_active => self.foreground.into_rgba(),
+            Element::Text => self.dim_foreground.into_rgba(),
         }
     }
 
-    fn close_button_color(&self, _status: ButtonState) -> [u8; 4] {
-        [0x00, self.background.r, self.background.g, self.background.b]
-    }
-
-    fn close_button_icon_color(&self, status: ButtonState) -> [u8; 4] {
-        self.color_icon_color(self.hovered_close_icon, status)
-    }
-
-    fn maximize_button_color(&self, _status: ButtonState) -> [u8; 4] {
-        [0x00, self.background.r, self.background.g, self.background.b]
-    }
-
-    fn maximize_button_icon_color(&self, status: ButtonState) -> [u8; 4] {
-        self.color_icon_color(self.hovered_maximize_icon, status)
+    fn button_color(
+        &self,
+        button: Button,
+        state: ButtonState,
+        foreground: bool,
+        window_active: bool,
+    ) -> ARGBColor {
+        match (foreground, button) {
+            (false, _) => ARGBColor { a: 0x00, r: 0x00, g: 0x00, b: 0x00 },
+            (_, Button::Minimize) => self
+                .button_foreground_color(self.hovered_minimize_icon, state, window_active)
+                .into_rgba(),
+            (_, Button::Maximize) => self
+                .button_foreground_color(self.hovered_maximize_icon, state, window_active)
+                .into_rgba(),
+            (_, Button::Close) => self
+                .button_foreground_color(self.hovered_close_icon, state, window_active)
+                .into_rgba(),
+        }
     }
+}
 
-    fn minimize_button_color(&self, _status: ButtonState) -> [u8; 4] {
-        [0x00, self.background.r, self.background.g, self.background.b]
-    }
+trait IntoARGBColor {
+    fn into_rgba(self) -> ARGBColor;
+}
```

Comment:
I'd assume you solved this with a separate trait since you can't implement `From<Rgb>` or `Into<ARGBColor` because you own neither one nor the other?

Reply:
Yes.